### PR TITLE
Vote Meeting Times JavaScript DAO

### PR DESCRIPTION
**Pull request by anna-giang**
This DAO will be used to access the servlet at `/vote-meeting-time` endpoint to increment the votes and add to the voters list of the meeting time voted for.

### CODE REVIEW

File: GoMeet/src/main/webapp/js/dao-utils.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class DAOUtils {
```
Comment (theresenieva) [inline GoMeet/src/main/webapp/js/dao-utils.js]: Moving this to a utils class was a smart idea :)

**Review by theresenieva — APPROVED** [review]
lgtm :)

File: GoMeet/src/main/webapp/js/meeting-time-dao.js [hunk 2]
```diff
@@ -38,9 +21,10 @@ class MeetingTimeDAO {
       throw new Error(INVALID_PARAM_TYPE);
     }
 
-    let urlString = MeetingTimeDAO.url({'meetingTimeId': meetingTimeId});
-    let results = await fetch(urlString).then((results) => results.json());
-    return results;
+    let urlString = DAOUtils.url(MeetingTimeDAO.endpoint, {'meetingTimeId': meetingTimeId});
+    let results = await fetch(urlString);
+    let resultsJson = await results.json();
```
Comment (vivianho18) [inline GoMeet/src/main/webapp/js/meeting-time-dao.js]: Why did you change this from the arrow function style from before?

File: GoMeet/src/main/webapp/js/meeting-time-dao.js [hunk 3]
```diff
@@ -76,11 +60,12 @@ class MeetingTimeDAO {
...
     // encode the data to be sent in the query string
-    let urlString = MeetingTimeDAO.url({'datetime': datetimeStr}); 
+    let urlString = DAOUtils.url(MeetingTimeDAO.endpoint, {'datetime': datetimeStr}); 
     let responseInit = {
       method: 'POST'
     }
-    let response = await fetch(urlString, responseInit).then((response) => response.json());
-    return response;
+    let response = await fetch(urlString, responseInit);
+    let responseJson = await response.json();
```
Comment (vivianho18) [inline GoMeet/src/main/webapp/js/meeting-time-dao.js]: Same comment as above

File: GoMeet/src/main/webapp/js/vote-meeting-time-dao.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    // encode the data to be sent in the query string
+    let urlString = DAOUtils.url(VoteMeetingTimeDAO.endpoint, {
+      'meetingTimeId': meetingTimeId,
+      'voters': voter
+    }); 
+    let responseInit = {
+      method: 'POST'
+    }
+    let response = await fetch(urlString, responseInit);
+    let responseJson = await response.json();
```
Comment (vivianho18) [inline GoMeet/src/main/webapp/js/vote-meeting-time-dao.js]: Same as above

File: GoMeet/src/main/webapp/test/spec/spec-vote-meeting-time-dao.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  const MEETING_TIME_ID = 'abc123def456';
+  const NON_EXISTENT_ID = 'non-existent-id';
+  const VOTER = 'John Citizen';
+  const QUERY_STRING = 'meetingTimeId=' 
+      + encodeURIComponent(MEETING_TIME_ID) 
+      + '&voters=' 
+      // Note that encodeURIComponenet encodes spaces to %20, 
+      // but URLSearchParam used in VoteMeetingTimeDAO class encodes 
+      // spaces to +. So need to replace that here
+      + encodeURIComponent(VOTER).replace(/%20/g, '+');
```
Comment (vivianho18) [inline GoMeet/src/main/webapp/test/spec/spec-vote-meeting-time-dao.js]: Why is it not `replace('%20', '+') ` ?

File: GoMeet/src/main/webapp/test/spec/spec-vote-meeting-time-dao.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    // Note: cannot use expect().toThrow, because the .toThrow executes before call to
+    // async voteMeetingTime is resolved.
+    let meetingTimeId = null;
+    let errorMessage;
+    try {
+      await VoteMeetingTimeDAO.voteMeetingTime(meetingTimeId, VOTER);
+    } catch (error) {
+      errorMessage = error.message;
+    }
+    expect(errorMessage).toEqual(INSUFFICIENT_REQUEST_PARAM);
```
Comment (vivianho18) [inline GoMeet/src/main/webapp/test/spec/spec-vote-meeting-time-dao.js]: Optional - since these lines are repeated in each of the error tests, it could be good to abstract that into a helper function. Something similar to what I've done at the bottom of this [test file ](https://github.com/googleinterns/step271-2020/pull/34/files#diff-75ddff44f177749d1205dd33040fe302f9350d3ab3a4b488fc3ae1983a45aae0)

**Review by vivianho18 — COMMENTED** [review]
Looking good :)

File: GoMeet/src/main/webapp/js/dao-utils.js (hunk 1 above)
Comment (anna-giang) [reply GoMeet/src/main/webapp/js/dao-utils.js]: Thankyou! 🤓

File: GoMeet/src/main/webapp/js/meeting-time-dao.js (hunk 2 above)
Comment (anna-giang) [reply GoMeet/src/main/webapp/js/meeting-time-dao.js]: I think when I rebased, commits were moved around, so the refactoring that I did was over written. I've written history, and changed it back now :)

File: GoMeet/src/main/webapp/test/spec/spec-vote-meeting-time-dao.js (hunk 5 above)
Comment (anna-giang) [reply GoMeet/src/main/webapp/test/spec/spec-vote-meeting-time-dao.js]: Using `replace('%20', '+')` will only replace the '%20' on the first occurrence, but using regex will replace all the %20 in the whole string, which is what I want.

File: GoMeet/src/main/webapp/test/spec/spec-vote-meeting-time-dao.js (hunk 6 above)
Comment (anna-giang) [reply GoMeet/src/main/webapp/test/spec/spec-vote-meeting-time-dao.js]: Yep, that's a good idea! I've changed it 😊

File: GoMeet/src/main/webapp/test/spec/spec-vote-meeting-time-dao.js (hunk 5 above)
Comment (anna-giang) [reply GoMeet/src/main/webapp/test/spec/spec-vote-meeting-time-dao.js]: But since we are not using user names to identify the user anymore, I've now updated the VOTER to be an email address, in which this replace statement is not used anymore.

File: GoMeet/src/main/webapp/js/vote-meeting-time-dao.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+   * @param {String} voter the name or identifier of the user voting for the
+   * MeetingTime
+   */
+  static async voteMeetingTime(meetingTimeId, voter) {
+    if (meetingTimeId === null || meetingTimeId === undefined) {
+      throw new Error(INSUFFICIENT_REQUEST_PARAM);
+    }
+    if (voter === null || voter === undefined) {
+      throw new Error(INSUFFICIENT_REQUEST_PARAM);
+    }
```
Comment (BipBopBoop) [inline GoMeet/src/main/webapp/js/vote-meeting-time-dao.js]: nit: these two conditionals can be combined into one.

File: GoMeet/src/main/webapp/js/vote-meeting-time-dao.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+    if (typeof meetingTimeId !== 'string' || typeof voter !== 'string') {
+      throw new Error(INVALID_PARAM_TYPE);
+    }
+
+    // encode the data to be sent in the query string
```
Comment (BipBopBoop) [inline GoMeet/src/main/webapp/js/vote-meeting-time-dao.js]: nit: Capitalize the 'e' and add a period at the end of the comment. Try and keep a consistent commenting style, elsewhere you capitalize the first word and end your comments with a period, I think that is good.

**Review by BipBopBoop — APPROVED** [review]
lgtm % nits

**Comment by anna-giang** [comment]
Rebasing on to main.
